doc/tools/docgram/doc_grammar.ml

```diff
@@ -0,0 +1,1349 @@
+open Coqpp_parser
+open Coqpp_ast
+
+let exit_code = ref 0
+
+let fprintf = Printf.fprintf
+
+let error s =
+  exit_code := 1;
+  Printf.eprintf "Error: ";
+  Printf.eprintf s
+
+let warn s =
+  Printf.eprintf "Warning: ";
+  Printf.eprintf s
+
+
+(* translated symbols *)
+
+type doc_symbol =
+| Sterm of string
+| Snterm of string
+| Slist1 of doc_symbol
+| Slist1sep of doc_symbol * doc_symbol
+| Slist0 of doc_symbol
+| Slist0sep of doc_symbol * doc_symbol
+| Sopt of doc_symbol
+| Sparen of doc_symbol list (* for GRAMMAR EXTEND *)
+| Sprod of doc_symbol list list (* for GRAMMAR EXTEND *)
+  (* editing operations *)
+| Sedit of string
+| Sedit2 of string * string
+
+(* nonterminals to rename or delete *)
+module StringMap = Map.Make(String)
+module NTMap = StringMap
+module StringSet = Set.Make(String)
+
+type gram = {
+  (* map from nonterminal name to a list of prods *)
+  (* each production is a list of doc_symbol *)
+  mutable map: doc_symbol list list NTMap.t;
+  (* specifies the order for nt's *)
+  mutable order: string list;
+}
```

Comment:
Is it an invariant that the `order` field is supposed to always be a total ordering of the keys from the `map` field? If so, please define functions adding and removing the non-terminals here and use them throughout the code, ideally encapsulating this type in a sealed module. Similarly, don't use `List.mem` on `order` when you could test membership on the set.

Also, rather than mutable fields, use a reference to a gram (or better no reference at all).

Reply:
> Is it an invariant that the `order` field is supposed to always be a total ordering of the keys from the `map` field?

Yes.  I created a module.  Not sure what I need to do to seal the module; I just followed an example from Stm.  I kept it as a ref field because otherwise I would have had to turn several routines inside out (i.e. the update is nested inside of several structures/loops).